specs/transport/car/carv2/index.md

```diff
@@ -0,0 +1,177 @@
+
+```
+| 11-byte fixed pragma | 40-byte header | optional padding | CARv1 data payload | optional padding | optional index payload
+```
+
+### Pragma
+
+The CARv2 version pragma (or "magic bytes") was chosen for compatibility with existing CARv1 parsers. CARv1 leads with a [DAG-CBOR](../../../codecs/dag-cbor/index.md) block, prefixed with a varint where the block contains the version and roots array:
+
+```ipldsch
+type CarV1Header struct {
+  version Int # 1
+  roots [&Any]
+}
+```
+
+To introduce a new version that existing parsers can safely reject as an *"unsupported version"* we must copy the minimal form of this header with a new version.
+
+Therefore, the CARv2 header is a fixed sequence of 11 bytes: **`0x0aa16776657273696f6e02`**.
+
+These bytes decode as follows: A leading `0x0a` which translates as a `uint(10)` (or `varint(10)`) indicating the length of the DAG-CBOR header block to follow. The remaining 10 bytes are a standard CBOR encoding of a map containing a `"version"` field with a value `2`. i.e.
+
+
+```ipldsch
+type CarV2Header struct {
+  version Int # 2
+}
+```
+
+The 10 bytes are parsed in CBOR as follows:
+
+```
+a1                                                # map(1)
+  67                                              #   string(7)
+    76657273696f6e                                #     "version"
+  02                                              #   uint(2)
+```
+
+Existing CARv1 parsers should safely read this pragma and reject the container as an unsupported version of the CAR format.
+
+This 11 byte string remains fixed and may be matched using a simple byte comparison and does not require a varint or CBOR decode since it does not vary for CARv2 containers.
+
+### Header
+
+Following the 11 byte pragma, the CARv2 is a fixed-length sequence of 40 bytes, broken into the following sections:
+
+1. **Characteristics**: A 128-bit (16-byte) bitfield used to describe certain features of the enclosed data.
+2. **Data offset**: A 64-bit (8-byte) unsigned little-endian integer indicating the byte-offset from the beginning of the container to the first byte of the CARv1 data payload.
+3. **Data length**: A 64-bit (8-byte) unsigned little-endian integer indicating the byte-length of the CARv1 data payload.
+4. **Index offset**: A 64-bit (8-byte) unsigned little-endian integer indicating the byte-offset from the beginning of the container to the first byte of the index payload.
+
+### Characteristics
+
+The characteristics bitfield contained within the CARv2 header may be used to indicate certain features of the specific CARv2 container.
+
+**TODO:** spit-balling here, these feel like feature-bloat so we should probably only add features as we use them but should be clear to leave enough space to encode additional things we care about without causing conflicts - i.e. it shouldn't be possible to flip bits to say that it's _both_ depth-first and breadth-first. Hence the suggestion to use easy integers where there's >2 options.
```

Comment:
Why not make this a CID:Data entry of dag-cbor instead of a custom data format? The first block being the header would be fine, it would future proof this a lot better too.

Reply:
Yeah, I'm not really excited about fact that we're yet again inventing another custom format when we have good formats at hand already.

Maybe the main reason to choose this was has to do with determinism of location, which is a problem we have in CARv1 now since the header is a flexible size and coming back and updating it after you've written the data is difficult.

But it's true that these fields are all fixed-length and don't have the problem of containing a flexible-length CIDs array like the CARv1 header so this could easily be just a CBOR array. If we're future proofing we'd need to add a varint length prefix too so we know how big the header is. I don't know if we need the CID, CARv1 doesn't bother doing that with the header block (which does add inconsistency to the format which is slightly annoying, but it's precedent).

In IPLD Schema it could be:

```
type CARv2Header struct {
  characteristics Bytes
  dataOffset Int
  dataSize Int
  indexOffset Int
} representation tuple
```

Reply:
```
type CARv2Header struct {
  dataOffset Int
  dataSize Int
  indexOffset Int
  characteristics Bytes
} representation tuple
```

With `characteristics` last might make more sense. But as long as we're going with dag-cbor we really should consider this field and whether we should just add explicit entries for it. `deduped Bool`, `orderingMethod Int`.